lib/ABI/Definition.cpp

```diff
@@ -0,0 +1,249 @@
+    const auto Iterator = llvm::find(GPRs, D.ReturnValueLocationRegister());
+    if (Iterator != GPRs.end() && Iterator != GPRs.begin())
+      return false;
+  }
+
+  return true;
+}
+
+using RegisterVector = std::vector<model::Register::Values>;
+
+template<typename RegisterType>
+bool compatibilityHelper(const SortedVector<RegisterType> &UsedRegisters,
+                         const RegisterVector &AllowedRegisters,
+                         model::Architecture::Values Architecture) {
+  for (const RegisterType &Register : UsedRegisters) {
+    // Verify the architecture of used registers.
+    if (model::Register::getArchitecture(Register.Location()) != Architecture)
+      revng_abort();
+  }
+
+  // Verify that every used register is also allowed.
+  for (const RegisterType &Register : UsedRegisters)
+    if (llvm::count(AllowedRegisters, Register.Location()) != 1)
+      return false;
+
+  return true;
+}
+
+namespace abi {
+
+bool Definition::verify() const {
+  if (ABI() == model::ABI::Invalid)
+    return false;
+
+  const auto Architecture = model::ABI::getRegisterArchitecture(ABI());
+  if (!verifyRegisters(GeneralPurposeArgumentRegisters(), Architecture))
+    return false;
+  if (!verifyRegisters(GeneralPurposeReturnValueRegisters(), Architecture))
+    return false;
+  if (!verifyRegisters(VectorArgumentRegisters(), Architecture))
+    return false;
+  if (!verifyRegisters(VectorReturnValueRegisters(), Architecture))
+    return false;
+  if (!verifyRegisters(CalleeSavedRegisters(), Architecture))
+    return false;
+
+  if (!verifyReturnValueLocation(*this))
+    return false;
+
+  // TODO: extend to verify `Types` once it's more substantial,
+  //       i.e. holds vector argument information.
+
+  return true;
+}
+
+template<typename... Ts>
+using SV = SortedVector<Ts...>;
+
+template<typename AR, typename RVR>
+bool Definition::checkCompatibility(const SV<AR> &ArgumentRegisters,
+                                    const SV<RVR> &ReturnValueRegisters) const {
+  revng_assert(verify());
+
+  const auto Architecture = model::ABI::getRegisterArchitecture(ABI());
+  if (!compatibilityHelper(ArgumentRegisters,
+                           GeneralPurposeArgumentRegisters(),
+                           Architecture))
+    return false;
+  if (!compatibilityHelper(ReturnValueRegisters,
+                           GeneralPurposeReturnValueRegisters(),
+                           Architecture))
+    return false;
+
+  return true;
+}
+
+using D = Definition;
+using TR = model::TypedRegister;
+using NTR = model::NamedTypedRegister;
+
+// For now, only enable the common check.
+// TODO: define more of these specializations if the need arises.
+template bool
+D::checkCompatibility<NTR, TR>(const SV<NTR> &ArgumentRegisters,
+                               const SV<TR> &ReturnValueRegisters) const;
+
+static llvm::StringRef translateABIName(model::ABI::Values ABI) {
+  switch (ABI) {
+  case model::ABI::SystemV_x86_64:
```

Comment:
There's an ongoing effort project wide to reduce the number of names that refer to the same thing.
It'd be much more straightforward if the file name was the same as the ABI.

Reply:
It probably would. The thing is, I feel like the names we have for the ABIs now are not ideal. Past me is definitely the culprit.

I suggest we let this stay for a bit until we move the `model::ABI::Values` onto the runtime (hopefully soon) at which point we can choose whichever names we want which will then be the only way we identify a specific ABI in revng.

What do you think?

Reply:
I think that it's better to have one bad name than one bad and another name on top of it.
Let's rename the files as the enum entries.

Reply:
Gotcha.

Reply:
However sad it makes me, here it is: https://github.com/revng/revng/pull/232/commits/605d703a29fd6ccc36bfb4e1af0603c667a59ac6.